test/metabase/http_client.clj

```diff
@@ -219,8 +235,101 @@
+  (with-open [os (java.io.ByteArrayOutputStream.)]
+    (let [f             (.f streaming-response)
+          canceled-chan (a/promise-chan)]
+      (f os canceled-chan)
+      (cond-> (.toByteArray os)
+        (some #(re-find % content-type) [#"json" #"text"])
+        (String. "UTF-8")))))
+
+(defn- coerce-mock-req-body
+  [resp]
+  (update resp :body
+          (fn [body]
+            (cond
+             ;; read the text respone
+             (instance? InputStream body)
+             (with-open [r (io/reader body)]
+               (slurp r))
+
+             ;; read byte array stuffs like image
+             (instance? (Class/forName "[B") body)
+             (String. body "UTF-8")
+
+             ;; Most APIs that execute a request returns a streaming response
+             (instance? StreamingResponse body)
+             (read-streaming-response body (get-in resp [:headers "Content-Type"]))
+
+             :else
+             body))))
+
+(defn- build-mock-request
+  [{:keys [query-parameters url credentials http-body method request-options]}]
+  (let [http-body   (test-runner.assert-exprs/derecordize http-body)
+        query-parameters (merge
+                          query-parameters
+                          ;; sometimes we include the param in the URL(even though we shouldn't)
+                          (reduce (fn [acc query]
+                                    (let [[k v] (str/split query #"=")]
+                                      (assoc acc k v)))
+                                  {}
+                                  (some-> (java.net.URI. url)
+                                          .getRawQuery
+                                          (str/split #"&"))))
+        url             (cond->> (first (str/split url #"\?")) ;; strip out the query param parts if any
+                          (not= (first url) \/)
+                          (str "/"))]
+    (-> (merge
+         {:accept         "json"
+          :headers        {"content-type" "application/json"
+                           @#'mw.session/metabase-session-header (when credentials
+                                                                   (if (map? credentials)
+                                                                     (authenticate credentials)
+                                                                     credentials))}
+          :query-string   (build-query-string query-parameters)
+          :request-method method
+          :uri            (str *url-prefix* url)}
+         (when (seq http-body)
+           {:body http-body})
+         request-options)
+        (m/update-existing :body (fn [http-body]
+                                   (java.io.ByteArrayInputStream.
+                                    (.getBytes (if (string? http-body)
+                                                 http-body
+                                                 (json/generate-string http-body)))))))))
+
+
+(schema/defn ^:private -mock-client
```

Comment:
Please don't use `schema` any more, use Malli going forward

Reply:
Converted this ns to Malli here https://github.com/metabase/metabase/pull/32070/commits/d6368c39df2556f5c1c4d47badc4ee9fba267021